In [21]:
#Task 1.1
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import OneHotEncoder

hotel_reservations = pd.read_csv('/kaggle/input/lab4ds/Hotel Reservations.csv', index_col='Booking_ID')

In [22]:
#Task 1.2
print("Shape of the dataset:", hotel_reservations.shape)
print("Number of samples:", len(hotel_reservations))
print("Number of features:", len(hotel_reservations.columns))
cat_cols = hotel_reservations.select_dtypes(include=['object']).columns.tolist()
print("Number of categorical features:", len(cat_cols))
print("Distribution of booking_status:")
print(hotel_reservations['booking_status'].value_counts())

Shape of the dataset: (36275, 18)
Number of samples: 36275
Number of features: 18
Number of categorical features: 4
Distribution of booking_status:
Not_Canceled    24390
Canceled        11885
Name: booking_status, dtype: int64


In [23]:
#Task 1.3
X = hotel_reservations.iloc[:, :-1]
y = hotel_reservations.loc[:, 'booking_status']

In [24]:
#Task 1.4
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

In [25]:
#Task 1.5
encoder = OneHotEncoder(handle_unknown='ignore')
X_train_encoded = encoder.fit_transform(X_train.select_dtypes(include='object'))
X_test_encoded = encoder.transform(X_test.select_dtypes(include='object'))
dtc = DecisionTreeClassifier(random_state=100)
dtc.fit(X_train_encoded, y_train)
y_pred = dtc.predict(X_test_encoded)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy of DecisionTree Classifier: {accuracy:.4f}")

Accuracy of DecisionTree Classifier: 0.6805


In [26]:
#Task 1.6
X_numeric = X.select_dtypes(exclude='object')

In [27]:
#Task 1.7
X_train_numeric, X_test_numeric, y_train_numeric, y_test_numeric = train_test_split(X_numeric, y, test_size=0.2, random_state=100)
dtc_numeric = DecisionTreeClassifier(random_state=100)
dtc_numeric.fit(X_train_numeric, y_train_numeric)
y_pred_numeric = dtc_numeric.predict(X_test_numeric)
accuracy_numeric = accuracy_score(y_test_numeric, y_pred_numeric)
print(f"Accuracy of DecisionTree Classifier using only numeric features: {accuracy_numeric:.4f}")

Accuracy of DecisionTree Classifier using only numeric features: 0.8582


#Task 1.7


Based on the output, the accuracy of the model with all features is:0.6805

Accuracy of DecisionTree Classifier with All Features: 0.6805
The accuracy of the model with only numeric features is:0.8582

Accuracy of DecisionTree Classifier with Numeric Features Only: 0.8582
The accuracy of the model with all features is better than the accuracy of the model with only numeric features.

In [29]:
#Task 1.8
from sklearn.model_selection import GridSearchCV

param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [2, 4, 6, 8, 10],
    'max_features': ['sqrt', 'log2', None],
    'splitter': ['best', 'random']
}

dtc = DecisionTreeClassifier(random_state=100)

gs_dtc = GridSearchCV(dtc, param_grid, cv=5)
gs_dtc.fit(X_train_encoded, y_train)


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=100),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [2, 4, 6, 8, 10],
                         'max_features': ['sqrt', 'log2', None],
                         'splitter': ['best', 'random']})

In [30]:
#Task 1.9
print(f"Best hyperparameters: {gs_dtc.best_params_}")

Best hyperparameters: {'criterion': 'gini', 'max_depth': 4, 'max_features': 'sqrt', 'splitter': 'random'}


In [31]:
#Task 1.10
dtc_best = DecisionTreeClassifier(**gs_dtc.best_params_, random_state=100)
dtc_best.fit(X_train_encoded, y_train)
y_pred_best = dtc_best.predict(X_test_encoded)
accuracy_best = accuracy_score(y_test, y_pred_best)
print(f"Accuracy of DecisionTree Classifier with best hyperparameters: {accuracy_best:.4f}")

Accuracy of DecisionTree Classifier with best hyperparameters: 0.6805


#Task 1.11


Accuracy of DecisionTree Classifier with Best Hyperparameters: 0.6805

This means that the accuracy of the new model is 0.6805,
which is slightly lower than the accuracy of the original model without hyperparameter tuning.
Therefore, it can be concluded that the new model does not improve on the original model

In [33]:
#Task 2.1
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer

hotel_reservations = pd.read_csv('/kaggle/input/lab4ds/Hotel Reservations.csv', index_col='Booking_ID')

column_transformer = make_column_transformer((OneHotEncoder(), ['type_of_meal_plan', 'room_type_reserved', 'market_segment_type']), remainder='passthrough')
X_transformed = column_transformer.fit_transform(hotel_reservations.iloc[:, :-1])
X_transformed = pd.DataFrame(data=X_transformed)

y = hotel_reservations.loc[:, 'booking_status']

X_train, X_test, y_train, y_test = train_test_split(X_transformed, y, test_size=0.2, random_state=100)

In [34]:
#Task 2.2
dtc = DecisionTreeClassifier(random_state=100)
dtc.fit(X_train, y_train)
y_pred = dtc.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy of DecisionTree Classifier with All Features: {accuracy:.4f}")

Accuracy of DecisionTree Classifier with All Features: 0.8692


#Task 2.2

Accuracy of DecisionTree Classifier with All Features: 0.8692
The accuracy of the model using all features, including categorical features,
is very high, suggesting that the categorical features provide useful information for the model to make accurate predictions.

Student name: Khalid Nimri

Student ID: 2140145